### ISSUE TITLE
resource.raccess.public is set without setting isPublic AVU, so iRODS and Django disagree on public status

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#2071]
@hyi @pkdash @mjstealey @dtarb
In issue #2031 @hyi asked why the `isPublic` AVU doesn't always match the `ResourceAccess.public` flag according to `check_irods_files`. I found out why just now. 

In the REST API, the setting of the `isPublic` AVU is carefully modified to always match the setting of `ResourceAccess.public` using the private routine `hs_core/views/__init__.py:_set_resource_sharing_status`. 

However, there are several places in the code (`hs_core/HydroShare/resource.py`, `hs_core/views/utils.py`, and `hs_core/views/__init__.py` in which _this is not used_, `ResourceAccess.public` is modified, and the AVU is **not** set to match `ResourceAccess.public`. There are several places where `Resource.public` is set to `False`, and one place where `ResourceAccess.public` is set to `True` and HYRAX updates are not scheduled as documented in `_set_resource_sharing_status`.  

Will attempt a hotfix for 1.10.0.

**Comment by hyi** [comment]
@alvacouch Thanks for finding the issue and for trying to make a fix to it. I think it'd be good to make the private method ```_set_resource_sharing_status``` public so that it can be the single point of truth for changing resource sharing flag.

**Comment by alvacouch** [comment]
@hyi I have taken a different approach. There are routines `BaseResource.set_public(Bool)` and `BaseResource.set_discoverable(Bool)` that understand the logic and make corrections. 
I left `_set_resource_sharing_status` alone, as it is for REST, and I need something for the native API which is where public is being changed. Will post PR soon.

**Comment by hyi** [comment]
@alvacouch Sounds good. I'll keep an eye out on the PR then.

**Comment by alvacouch** [comment]
@dtarb @horsburgh I noticed in fixing this issue that there are a number of places around the code where **we allow a resource to be public and not discoverable.** Is this intentional or an oversight? 

It would seem that making something public would implicitly make it discoverable, and that making something not discoverable would implicitly make it not public. 

Is there a use case for resources that are public but never show up on the discovery interface?

**Comment by dtarb** [comment]
I think this is an oversight.  We have written in our access control policy "Public implies Discoverable, but not the other way around" and I do not have any recollection of ever deciding anything else.  I suppose there could  be a use case of security by obscurity where we allow this, but I do not think we should do this at this stage, and it would require us to change the UI (again something I do not want to advocate at this stage).

**Comment by horsburgh** [comment]
Agreed. I say public = discoverable.

**Comment by hyi** [comment]
@alvacouch I agree with @dtarb and @horsburgh, but want to point out this just in case. If you are looking at public flag and discoverable flag in the code base, as I remembered, we use ```public=False and discoverable=True``` to imply a resource is discoverable, and if public is true, then the resource is public without even looking at discoverable flag. Looping in @pkdash in case he has more to say regarding this.

**Comment by alvacouch** [comment]
@dtarb @hyi @horsburgh @pkdash 
I have "discovered" (very bad pun intended) that updating of `discoverable` and `public` in response to resource metadata changes is very, very inconsistent.  Hence, so is updating of `isPublic`.

In `add_metadata`, it's updated **only if deleting all subjects,** (???) and then, if metadata still validates after doing that, , `discoverable` is enabled even if it's currently turned off!  (!!!)  (What were we thinking?) 

In other metadata routines and file edits (`update_metadata`, `move_file_or_folder`, `delete_folder`, ...) , requirements for `public` or `discoverable` aren't checked and resources are allowed to be `discoverable` after changes make absence of metadata or files insufficient for discovery.  

In summary, the majority of the code: 
1. Enforces metadata and file requirements when setting something discoverable or public. 
2. Does not continue to enforce these standards when something public or discoverable gets changed. 

I'm a bit worried about "fixing" this in 1.10.0, because there's no time to see whether this will be accurately reflected in the UI.  So I am peppering the code with TODOs.

**Comment by alvacouch** [comment]
Just to expedite 1.10.0, I suggest that we:

1. **remove the obvious silliness** in `add_metadata`. 
2. **leave all of the native API not validating** (like it is now). Changes within the UI will never set `discoverable` to False even if that is quite well deserved.  
3. leave the REST API validating as it is now.

And then sort out the mess in 1.11.0

**Comment by dtarb** [comment]
I am OK with this approach.

**Comment by pkdash** [comment]
@alvacouch  Looking at the silliness of the ```add_metadata``` (probably I wrote the code to fix an issue with editing of resource level keywords) the code that enables 'discoverable' is never executed. 

The fix you have proposed for 1.10 sounds good.

**Comment by pkdash** [comment]
@alvacouch 

> In other metadata routines and file edits (update_metadata, move_file_or_folder, delete_folder, ...) , requirements for public or discoverable aren't checked and resources are allowed to be discoverable after changes make absence of metadata or files insufficient for discovery.

There is no need to set public/discoverable in case of metadata update (since only existing metadata getting is getting updated). Only in the case of resource title being updated, we check to see if the resource public to be reset.

delete_metadata_element() view function should have been checking for possible resting of these flags. The reason it was not done that the UI does not allow to delete any required metadata elements. The only required metadata elements that can be deleted is the keyword which is part of the add_metadata_element() view function where we check (silly code) for resetting these flags. 

In case of delete folder, which calls delete_resource_file,  public/discoverable is reset if needed. 

If we are moving file or folder I think there is no need to check or reset these flags.

**Comment by pkdash** [comment]
I am not sure how the discovery interface filters resources in terms of using the discoverable/public flags. From the resource landing page point of view, for anonymous user we check the flag public to decide if content files be displayed or not. 

REST api for accessing metadata xml file the authorization is successful if either the resource is public or discoverable.

**Comment by alvacouch** [comment]
@pkdash Discovery displays something if it is either public or discoverable. 

The approach above presumes a lot about the validation process (if only subject and title are flagged as being necessary to check). If someone decides to validate more metadata, this will have to be changed.

**Comment by alvacouch** [comment]
@pkdash I also note that validation of logical file metadata only occurs during changes to logical files, and **is not checked when one tries to publish a resource.** I am changing that.

**Comment by alvacouch** [comment]
Merged #2084 to repair this for Release Candidate 1.10.0

**Comment by alvacouch** [comment]
Theoretically, this has been addressed.
